### PR TITLE
battery_stateful

### PR DESCRIPTION
**Pull request by dguittet**
Overall:
 - New `cmod_battery_stateful`
     - Useage example at this [test](https://github.com/NREL/ssc/blob/batt_refactor/test/ssc_test/cmod_battery_stateful_test.cpp)
 - `core.h` modifications for `ssc_module_stateful_create` in sscapi
 - Split `lib_battery.t` into component parts
 - Separate all class variables into `params`, which are fixed for the model simulation, and `state`, which change according to how the model is executed
     - `params` and `state` ownership is shared across all components of battery model using `shared_ptr`
 - Enabled `set_state` and `get_state` for all models
 - Consolidated and organized copying and assignment functions
 - New constructors to create models from `params`
 - Use `unique_ptr` for component models in `battery_t`
 - Removed all leading underscore for class members. 
 - Add logging and overload << operator
 - Add tests for all models
 - Compute_module functions to exchange data between `params` and `state` structs with ssc `var_table`

capacity_t:
 - Fixed `calculate_max_discharge_w` and `calculate_max_charge_w` bugs for vanadium redox and table models

thermal_t:
 - Switched to analytic solution of heat transfer equation
 - Replaced height, length & width with surface_area
 - Changed thermal parameters in all tests: resistance, Cp, h and surface area
 - Switched to Celsius
 - Added heat_dissipated to state
 - Added `batt_temperature` to `cmod_battery` output

lifetime_t:
 - Moved replacement functionality into `battery_t`

Changes to tests:
 - Many changes due to different thermal behavior
 - Changes to FOM dispatch tests due to sensitivity of battery target powers to the cycle cost

fix #291-- calendar degradation across 25 years is same regardless of timestep for current thermal defaults

Remaining items:
 - Losses are not applied
 - Thermal defaults need to be updated

@sjanzou Steve, I've made some changes to `core.h`, will you please take a look at `DEFINE_STATEFUL_MODULE_ENTRY`?

### CODE REVIEW

**Comment by brtietz** [comment]
I tried to build this on Windows, and got:

Severity	Code	Description	Project	File	Line	Suppression State
Error	C2039	'bind': is not a member of 'std'	shared	C:\Users\bmirletz\source\repos\sam_dev\ssc\shared\lib_battery_voltage.cpp	405	

and

Severity	Code	Description	Project	File	Line	Suppression State
Error	C2039	'function': is not a member of 'std'	shared	C:\Users\bmirletz\source\repos\sam_dev\ssc\shared\lib_battery_voltage.cpp	402	

Is there a missing include? (#include <functional> ?)

**Comment by dguittet** [comment]
Oh thanks for catching that. It should be

#include <functional>

File: shared/CMakeLists.txt [hunk 1]
```diff
@@ -34,10 +34,16 @@ set(SHARED_SRC
 		lib_aoi.h
 		lib_battery.cpp
 		lib_battery.h
+        lib_battery_capacity.cpp
```
Comment (brtietz) [inline shared/CMakeLists.txt:37]: Looks like there are some tabs/spaces issues here. Is it worth fixing the whole list so the indentation isn't wonky?

File: ssc/cmod_battery_stateful.cpp [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+
+#include <lib_battery_capacity.h>
+#include "common.h"
+#include "vartab.h"
+#include "core.h"
+#include "cmod_battery_stateful.h"
+
+var_info vtab_battery_stateful_inputs[] = {
+        /*   VARTYPE           DATATYPE         NAME                                            LABEL                                                   UNITS      META                   GROUP           REQUIRED_IF                 CONSTRAINTS                      UI_HINTS*/
+        { SSC_INPUT,        SSC_NUMBER,      "control_mode",                               "Control using current (0) or power (1)",                  "hr",      "",   "Controls",       "*",                           "",                              "" },
```
Comment (brtietz) [inline ssc/cmod_battery_stateful.cpp:31]: Units should be 0/1

File: test/ssc_test/cmod_battery_stateful_test.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    ssc_data_get_number(data, "V", &V);
+    ssc_data_get_number(data, "SOC", &SOC);
+    EXPECT_EQ(last_idx, 2);
+    EXPECT_NEAR(current, 1, 1e-2);
+    EXPECT_NEAR(P, 0.457, 1e-2);
+    EXPECT_NEAR(V, 457.93, 1e-2);
+    EXPECT_NEAR(SOC, 41.79, 1e-2);
+}
+
+TEST_F(CMBatteryStatefulIntegration_cmod_battery_stateful, compareSequential) {
```
Comment (brtietz) [inline test/ssc_test/cmod_battery_stateful_test.cpp:52]: The runtime comparison is definitely useful, but should this test assert anything? One thing I could think of would be making sure the "without reading state" runtime is lower.

File: test/shared_test/lib_battery_test.cpp [hunk 4]
```diff
@@ -79,17 +79,17 @@ TEST_F(BatteryTest, AugmentCapacity)
...
-	batteries.push_back(new battery_t(dtHour, chemistry));
-	batteries[2]->initialize(capacityModel, voltageModel, lifetimeModel, thermalModel, lossModel);
-	batteries[2]->lifetime_model()->set_replacement_option(battery_t::REPLACE_BY_SCHEDULE);
+    batteries.push_back(new battery_t(*batteryModel));
+
+	batteries.push_back(new battery_t(*batteryModel));
+	batteries[1]->setupReplacements(replacement_schedule, augmentation_percent);
+
+	batteries.push_back(new battery_t(*batteryModel));
+	batteries[2]->setupReplacements(replacement_schedule, augmentation_percent);
```
Comment (brtietz) [inline test/shared_test/lib_battery_test.cpp:92]: The augment capacity test does not assert anything.

File: test/shared_test/lib_battery_model_test.h [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+}
+
+static void compareState(const std::shared_ptr<thermal_state>& tested_state, const std::shared_ptr<thermal_state>& expected_state, const std::string& msg){
+    compareState(*tested_state, *expected_state, msg);
+}
+
+class lib_battery_thermal_test : public ::testing::Test
+{
+protected:
+//    std::unique_ptr<battery_capacity_interface> new_cap;
```
Comment (brtietz) [inline test/shared_test/lib_battery_model_test.h:48]: Now that these classes are split out, can these commented out lines be removed?

File: test/shared_test/lib_battery_model_test.cpp [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+        double output_power = 0.;
+        while(voltageModel->calculate_max_discharge_w(capacityModel->q0(), capacityModel->qmax(), 0, nullptr) > 0 ){
+            double output_current = current;
+            capacityModel->updateCapacity(output_current, 1);
+            voltageModel->updateVoltage(capacityModel->q0(), capacityModel->qmax(), capacityModel->I(), 0, 1);
+            output_power += capacityModel->I() * voltageModel->battery_voltage();
+            n_t += 1;
+        }
+//        printf("current %f, eff %f, n %d\n", current, -output_power/input_power, n_t);
+        current += fabs(max_current) / 100.;
```
Comment (brtietz) [inline test/shared_test/lib_battery_model_test.cpp:368]: Test has no assertions.

File: test/shared_test/lib_battery_model_test.cpp [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+        while(voltageModel->calculate_max_discharge_w(capacityModel->q0(), capacityModel->qmax(), 0, nullptr) > 0 ){
+            double output_current = current;
+            capacityModel->updateCapacity(output_current, 1);
+            voltageModel->updateVoltage(capacityModel->q0(), capacityModel->qmax(), capacityModel->I(), 0, 1);
+            output_power += capacityModel->I() * voltageModel->battery_voltage();
+            n_t += 1;
+
+        }
+//        printf("current %f, eff %f, n %d\n", current, -output_power/input_power, n_t);
+        current += fabs(max_current) / 100.;
```
Comment (brtietz) [inline test/shared_test/lib_battery_model_test.cpp:326]: Test has no assertions.

File: test/shared_test/lib_battery_model_test.cpp [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+    log << *params.thermal << "\n\n";
+
+    log << *state.losses << "\n";
+    log << *params.losses << "\n\n";
+
+    log << *state.replacement << "\n";
+    log << *params.replacement << "\n\n";
+
+    log << state << "\n";
+    log << params << "\n";
```
Comment (brtietz) [inline test/shared_test/lib_battery_model_test.cpp:288]: Test has no assertions.

File: test/shared_test/lib_battery_model_test.cpp [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+    auto bat = battery_t(params);
+
+    double current = 10;
+    batteryModel->run(0, current);
+    current = 10;
+    bat.run(0, current);
+
+    logger log(std::cout);
+    log << *params << "\n" << bat.get_params() << "\n\n";
+    log << batteryModel->get_state() << "\n" << bat.get_state() << "\n";
```
Comment (brtietz) [inline test/shared_test/lib_battery_model_test.cpp:260]: Test has no assertions.

File: test/shared_test/lib_battery_model_test.cpp [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+    double I = 10;
+    Battery->run(0, I);
+
+    state = batteryModel->get_state();
+    auto cap_state2 = state.capacity;
+    auto volt_state2 = state.voltage;
+
+    state = Battery->get_state();
+    auto cap_state3 = state.capacity;
+    auto volt_state3 = state.voltage;
```
Comment (brtietz) [inline test/shared_test/lib_battery_model_test.cpp:246]: Test has no assertions.

File: test/shared_test/lib_battery_lifetime_test.cpp [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+        EXPECT_EQ(state.cycle->q_relative_cycle, model->capacity_percent_cycle());
+        EXPECT_EQ(state.calendar->q_relative_calendar, model->capacity_percent_calendar());
+
+        idx ++;
+    }
+
+}
+
+
+//TEST_F(lib_battery_lifetime_test, ReplaceByCapacityTest){
```
Comment (brtietz) [inline test/shared_test/lib_battery_lifetime_test.cpp:261]: Can you add a comment with what needs to happen to get this test uncommented?

File: test/shared_test/lib_battery_capacity_test.h [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+    double q10 = 93.;
+    double t1 = 1;
+    double q1 = 60.;
+    double q20 = 100;
+public:
+
+    void SetUp() override {
+        old_cap = std::make_shared<capacity_kibam_t>(q20, t1, q1, q10,SOC_init,
+                SOC_max, SOC_min, dt_hour);
+//        double I = 1e-7;
```
Comment (brtietz) [inline test/shared_test/lib_battery_capacity_test.h:79]: Can these commented out lines be removed?

File: ssc/cmod_battery.cpp [hunk 13]
```diff
@@ -180,209 +178,209 @@ var_info vtab_battery_inputs[] = {
...
+        { SSC_OUTPUT,        SSC_ARRAY,      "batt_qmax_thermal",                          "Battery maximum charge at temperature",                  "Ah",       "",                     "Battery",       "",                           "",                              "" },
+        { SSC_OUTPUT,        SSC_ARRAY,      "batt_I",                                     "Battery current",                                        "A",        "",                     "Battery",       "",                           "",                              "" },
+        { SSC_OUTPUT,        SSC_ARRAY,      "batt_voltage_cell",                          "Battery cell voltage",                                   "V",        "",                     "Battery",       "",                           "",                              "" },
+        { SSC_OUTPUT,        SSC_ARRAY,      "batt_voltage",                               "Battery voltage",	                                     "V",        "",                     "Battery",       "",                           "",                              "" },
+
+        // Lifecycle related outputs
```
Comment (brtietz) [inline ssc/cmod_battery.cpp:202]: These comments are great, thank you!

File: ssc/cmod_battery.cpp [hunk 14]
```diff
@@ -1303,360 +1267,299 @@ battstor::battstor(const battstor& orig){
...
 
 extern var_info vtab_fuelcell_output[];
 
 void process_messages(shared_ptr<battstor> batt, compute_module* cm)
 {
     if (!batt)
         return;
-    message dispatch_messages = batt->dispatch_model->get_messages();
-    message thermal_messages = batt->thermal_model->get_messages();
+//    message dispatch_messages = batt->dispatch_model->get_messages();
```
Comment (brtietz) [inline ssc/cmod_battery.cpp:1498]: Can these be removed or do the messages need to be restored?

File: shared/CMakeLists.txt (hunk 1 above)
Comment (dguittet) [reply shared/CMakeLists.txt:37]: Oh that's strange as it didn't show up on my computer. Fixed!

File: test/shared_test/lib_battery_lifetime_test.cpp (hunk 11 above)
Comment (dguittet) [reply test/shared_test/lib_battery_lifetime_test.cpp:261]: I had moved the replacements out of lifetime and into the battery, so I cleaned up the test and moved it as part of lib_battery_test

File: test/ssc_test/cmod_battery_stateful_test.cpp (hunk 3 above)
Comment (dguittet) [reply test/ssc_test/cmod_battery_stateful_test.cpp:52]: Added thanks

File: ssc/cmod_battery.cpp (hunk 13 above)
Comment (dguittet) [reply ssc/cmod_battery.cpp:202]: Np!

File: ssc/cmod_battery.cpp (hunk 14 above)
Comment (dguittet) [reply ssc/cmod_battery.cpp:1498]: I removed the `message` class since I wanted to consolidate those into the logger. So will remove this.

**Comment by dguittet** [comment]
Thanks for the review! Made the requested changes and also added `operator==` overloading for some classes for easier comparison in tests. I wasn't comprehensive but such functionality could be added to other classes as necessary.

**Pull request merged**